### ISSUE TITLE
 memory leak caused by requestAudioFocus() on ReactExoplayerView

### ISSUE THREAD

**Issue by zhaozhe-ysy** [TheWidlarzGroup/react-native-video#1963]
# memory leak caused by requestAudioFocus() on ReactExoplayerView

## Platform

Which player are you experiencing the problem on:
* Android ExoPlayer

## Environment info

React native info output:

```bash
System:
    OS: macOS 10.15.3
    CPU: (4) x64 Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
    Memory: 18.35 MB / 8.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 13.0.1 - /usr/local/bin/node
    Yarn: 1.16.0 - /usr/local/bin/yarn
    npm: 6.12.0 - /usr/local/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 13.4, DriverKit 19.0, macOS 10.15, tvOS 13.4, watchOS 6.2
    Android SDK:
      API Levels: 17, 19, 21, 22, 23, 25, 26, 27, 28, 29
      Build Tools: 20.0.0, 21.1.2, 23.0.1, 23.0.2, 23.0.3, 25.0.0, 25.0.1, 25.0.2, 25.0.3, 26.0.2, 26.0.3, 27.0.2, 27.0.3, 28.0.0, 28.0.3, 29.0.0, 29.0.1
      System Images: android-24 | Google Play Intel x86 Atom
      Android NDK: 17.1.4828580
  IDEs:
    Android Studio: 3.6 AI-192.7142.36.36.6241897
    Xcode: 11.4/11E146 - /usr/bin/xcodebuild
  npmPackages:
    react: 16.9.0 => 16.9.0 
    react-native: 0.60.0 => 0.60.0

```

Library version: x.x.x

## Steps To Reproduce

1. render react-native-view
2. destroy react-native-view
3. repeat 1 and 2
4. ReactExoplayerView didn't release

## Expected behaviour

1. ReactExoplayerView can release

## Cause
when requestAudioFocus() was executed, the ReactExoplayerView instance will register in AudioManager and never release. If we didn't close app and repeat render view, it may cause memory leak.
## Solution
We found a solution to fix this issue, invoke audioManager.abandonAudioFocus() in the releasePlayer() as below.
```
    private void releasePlayer() {
        if (player != null) {
            updateResumePosition();
            player.release();
            player.setMetadataOutput(null);
            player = null;
            trackSelector = null;
        }
        progressHandler.removeMessages(SHOW_PROGRESS);
        themedReactContext.removeLifecycleEventListener(this);
        audioBecomingNoisyReceiver.removeListener();
        BANDWIDTH_METER.removeEventListener(this);
        audioManager.abandonAudioFocus(this);
    }
```

This is our solution, you can merge it or find other solution.

Thank you~

**Comment by Wolvereness** [comment]
I can confirm that there is an audio related leak and that this patch is insufficient to address it.
![image](https://user-images.githubusercontent.com/983111/90293671-408ab600-de4a-11ea-83bb-9187e7c3de2f.png)

I have a very long-running app, and the lion's share of the memory leak appears to be related to the sound management of the video player.

edit: I realized that the specific test I obtained this from didn't run with the patch. Rerunning.
edit2: Fixed the test and even after 4 hours it looks like the number of instances of players is close to the statistical average expected, i.e. none of them are getting GCed.

edit3: This patch appears to fix the leak, but my system doesn't use sound so it's probably not helpful for most:
```patch
Index: node_modules/react-native-video/android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- node_modules/react-native-video/android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java	(date 1597475492080)
+++ node_modules/react-native-video/android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java	(date 1597475492080)
@@ -3,7 +3,6 @@
 import android.annotation.SuppressLint;
 import android.app.Activity;
 import android.content.Context;
-import android.media.AudioManager;
 import android.net.Uri;
 import android.os.Handler;
 import android.os.Message;
@@ -15,8 +14,6 @@
 import android.widget.FrameLayout;
 
 import com.brentvatne.react.R;
-import com.brentvatne.receiver.AudioBecomingNoisyReceiver;
-import com.brentvatne.receiver.BecomingNoisyListener;
 import com.facebook.react.bridge.Arguments;
 import com.facebook.react.bridge.Dynamic;
 import com.facebook.react.bridge.LifecycleEventListener;
@@ -80,8 +77,6 @@
         LifecycleEventListener,
         ExoPlayer.EventListener,
         BandwidthMeter.EventListener,
-        BecomingNoisyListener,
-        AudioManager.OnAudioFocusChangeListener,
         MetadataRenderer.Output {
 
     private static final String TAG = "ReactExoplayerView";
@@ -149,8 +144,6 @@
 
     // React
     private final ThemedReactContext themedReactContext;
-    private final AudioManager audioManager;
-    private final AudioBecomingNoisyReceiver audioBecomingNoisyReceiver;
 
     private final Handler progressHandler = new Handler() {
         @Override
@@ -180,9 +173,7 @@
 
         createViews();
         
-        audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
         themedReactContext.addLifecycleEventListener(this);
-        audioBecomingNoisyReceiver = new AudioBecomingNoisyReceiver(themedReactContext);
 
         initializePlayer();
     }
@@ -357,7 +348,6 @@
                     player.addListener(self);
                     player.setMetadataOutput(self);
                     exoPlayerView.setPlayer(player);
-                    audioBecomingNoisyReceiver.setListener(self);
                     BANDWIDTH_METER.addEventListener(new Handler(), self);
                     setPlayWhenReady(!isPaused);
                     playerNeedsSource = true;
@@ -460,18 +450,11 @@
         }
         progressHandler.removeMessages(SHOW_PROGRESS);
         themedReactContext.removeLifecycleEventListener(this);
-        audioBecomingNoisyReceiver.removeListener();
         BANDWIDTH_METER.removeEventListener(this);
     }
 
     private boolean requestAudioFocus() {
-        if (disableFocus || srcUri == null) {
-            return true;
-        }
-        int result = audioManager.requestAudioFocus(this,
-                AudioManager.STREAM_MUSIC,
-                AudioManager.AUDIOFOCUS_GAIN);
-        return result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED;
+        return true;
     }
 
     private void setPlayWhenReady(boolean playWhenReady) {
@@ -533,7 +516,6 @@
             setFullscreen(false);
         }
         setKeepScreenOn(false);
-        audioManager.abandonAudioFocus(this);
     }
 
     private void updateResumePosition() {
@@ -558,39 +540,6 @@
         return DataSourceUtil.getDefaultDataSourceFactory(this.themedReactContext, useBandwidthMeter ? BANDWIDTH_METER : null, requestHeaders);
     }
 
-    // AudioManager.OnAudioFocusChangeListener implementation
-
-    @Override
-    public void onAudioFocusChange(int focusChange) {
-        switch (focusChange) {
-            case AudioManager.AUDIOFOCUS_LOSS:
-                eventEmitter.audioFocusChanged(false);
-                break;
-            case AudioManager.AUDIOFOCUS_GAIN:
-                eventEmitter.audioFocusChanged(true);
-                break;
-            default:
-                break;
-        }
-
-        if (player != null) {
-            if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK) {
-                // Lower the volume
-                player.setVolume(audioVolume * 0.8f);
-            } else if (focusChange == AudioManager.AUDIOFOCUS_GAIN) {
-                // Raise it back to normal
-                player.setVolume(audioVolume * 1);
-            }
-        }
-    }
-
-    // AudioBecomingNoisyListener implementation
-
-    @Override
-    public void onAudioBecomingNoisy() {
-        eventEmitter.audioBecomingNoisy();
-    }
-
     // ExoPlayer.EventListener implementation
 
     @Override
```

**Comment by KalumDog** [comment]
I face the problem too in **react-native-video@5.0.2**
any suggestion?

**Comment by Tang-xun** [comment]
It's my solution,  Maybe helpful to you 

Index: react-native-video/android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- node_modules/react-native-video/android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java	(revision e4478836f33397e8479bf24a51d83db5514b728f)
+++ node_modules/react-native-video/android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java	(revision a78bc362053c37f43f5058020079579041820914)
@@ -67,6 +67,7 @@
 import com.google.android.exoplayer2.ui.PlayerControlView;
 import com.google.android.exoplayer2.video.VideoFrameMetadataListener;
 
+import java.lang.ref.WeakReference;
 import java.net.CookieHandler;
 import java.net.CookieManager;
 import java.net.CookiePolicy;
@@ -82,7 +83,6 @@
         ExoPlayer.EventListener,
         BandwidthMeter.EventListener,
         BecomingNoisyListener,
-        AudioManager.OnAudioFocusChangeListener,
         VideoFrameMetadataListener,
         MetadataRenderer.Output {
 
@@ -102,7 +102,7 @@
     private PlayerControlView playerControlView;
     private View playPauseControlContainer;
     private Player.EventListener eventListener;
-    
+
     private Handler mainHandler;
     private ExoPlayerView exoPlayerView;
 
@@ -137,7 +137,7 @@
     private String audioTrackType;
     private Dynamic audioTrackValue;
     private String videoTrackType;
-    private Dynamic videoTrackValue;    
+    private Dynamic videoTrackValue;
     private ReadableArray audioTracks;
     private String textTrackType;
     private Dynamic textTrackValue;
@@ -175,13 +175,15 @@
         }
     };
 
+    private ExoOnAudioFocusChangeListener mOnAudioFocusChangeListener;
+
     public ReactExoplayerView(ThemedReactContext context) {
         super(context);
         this.themedReactContext = context;
         this.eventEmitter = new VideoEventEmitter(context);
 
         createViews();
-        
+
         audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
         themedReactContext.addLifecycleEventListener(this);
         audioBecomingNoisyReceiver = new AudioBecomingNoisyReceiver(themedReactContext);
@@ -466,14 +468,20 @@
         themedReactContext.removeLifecycleEventListener(this);
         audioBecomingNoisyReceiver.removeListener();
         BANDWIDTH_METER.removeEventListener(this);
-        audioManager.abandonAudioFocus(this);
+        audioManager.abandonAudioFocus(mOnAudioFocusChangeListener);
     }
 
     private boolean requestAudioFocus() {
         if (disableFocus || srcUri == null) {
             return true;
         }
-        int result = audioManager.requestAudioFocus(this,
+
+        if (mOnAudioFocusChangeListener == null) {
+            mOnAudioFocusChangeListener = new ExoOnAudioFocusChangeListener(player, eventEmitter);
+            updateAudioFocusChangeVolume();
+        }
+
+        int result = audioManager.requestAudioFocus(mOnAudioFocusChangeListener,
                 AudioManager.STREAM_MUSIC,
                 AudioManager.AUDIOFOCUS_GAIN);
         return result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED;
@@ -538,7 +546,7 @@
             setFullscreen(false);
         }
         setKeepScreenOn(false);
-        audioManager.abandonAudioFocus(this);
+        audioManager.abandonAudioFocus(mOnAudioFocusChangeListener);
     }
 
     private void updateResumePosition() {
@@ -565,29 +573,29 @@
 
     // AudioManager.OnAudioFocusChangeListener implementation
 
-    @Override
-    public void onAudioFocusChange(int focusChange) {
-        switch (focusChange) {
-            case AudioManager.AUDIOFOCUS_LOSS:
-                eventEmitter.audioFocusChanged(false);
-                break;
-            case AudioManager.AUDIOFOCUS_GAIN:
-                eventEmitter.audioFocusChanged(true);
-                break;
-            default:
-                break;
-        }
-
-        if (player != null) {
-            if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK) {
-                // Lower the volume
-                player.setVolume(audioVolume * 0.8f);
-            } else if (focusChange == AudioManager.AUDIOFOCUS_GAIN) {
-                // Raise it back to normal
-                player.setVolume(audioVolume * 1);
-            }
-        }
-    }
+//    @Override
+//    public void onAudioFocusChange(int focusChange) {
+//        switch (focusChange) {
+//            case AudioManager.AUDIOFOCUS_LOSS:
+//                eventEmitter.audioFocusChanged(false);
+//                break;
+//            case AudioManager.AUDIOFOCUS_GAIN:
+//                eventEmitter.audioFocusChanged(true);
+//                break;
+//            default:
+//                break;
+//        }
+//
+//        if (player != null) {
+//            if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK) {
+//                // Lower the volume
+//                player.setVolume(audioVolume * 0.8f);
+//            } else if (focusChange == AudioManager.AUDIOFOCUS_GAIN) {
+//                // Raise it back to normal
+//                player.setVolume(audioVolume * 1);
+//            }
+//        }
+//    }
 
     // AudioBecomingNoisyListener implementation
 
@@ -884,7 +892,7 @@
 
     public void setReportBandwidth(boolean reportBandwidth) {
         mReportBandwidth = reportBandwidth;
-    }   
+    }
 
     public void setRawSrc(final Uri uri, final String extension) {
         if (uri != null) {
@@ -1009,7 +1017,7 @@
             for (int j = 0; j < group.length; j++) {
                 tracks[j] = j;
             }
-        } 
+        }
 
         if (groupIndex == C.INDEX_UNSET) {
             trackSelector.setParameters(disableParameters);
@@ -1078,6 +1086,7 @@
         if (player != null) {
             player.setVolume(audioVolume);
         }
+        updateAudioFocusChangeVolume();
     }
 
 
@@ -1086,6 +1095,13 @@
         if (player != null) {
             player.setVolume(audioVolume);
         }
+        updateAudioFocusChangeVolume();
+    }
+
+    private void updateAudioFocusChangeVolume() {
+        if (mOnAudioFocusChangeListener != null) {
+            mOnAudioFocusChangeListener.setAudioVolume(audioVolume);
+        }
     }
 
     public void seekTo(long positionMs) {
@@ -1178,7 +1194,7 @@
 
     /**
      * Handling controls prop
-     * 
+     *
      * @param controls  Controls prop, if true enable controls, if false disable them
      */
     public void setControls(boolean controls) {
@@ -1198,4 +1214,53 @@
   public void onVideoFrameAboutToBeRendered(long presentationTimeUs, long releaseTimeNs, Format format) {
     this.presentationTimeUs = presentationTimeUs;
   }
+
+  // fix leaks for audio focus change listener
+  static class ExoOnAudioFocusChangeListener
+      implements AudioManager.OnAudioFocusChangeListener {
+
+
+      WeakReference<SimpleExoPlayer> refPlayer;
+      WeakReference<VideoEventEmitter> refEventEmitter;
+
+      float mAudioVolume;
+
+      public ExoOnAudioFocusChangeListener(SimpleExoPlayer player, VideoEventEmitter eventEmitter) {
+          refEventEmitter = new WeakReference<>(eventEmitter);
+          refPlayer = new WeakReference<>(player);
+      }
+
+      public void setAudioVolume(float audioVolume) {
+          mAudioVolume = audioVolume;
+      }
+
+      @Override
+      public void onAudioFocusChange(int focusChange) {
+          switch (focusChange) {
+              case AudioManager.AUDIOFOCUS_LOSS:
+                  if (refEventEmitter.get() != null) {
+                      refEventEmitter.get().audioFocusChanged(false);
+                  }
+                  break;
+              case AudioManager.AUDIOFOCUS_GAIN:
+                  if (refEventEmitter.get() != null) {
+                      refEventEmitter.get().audioFocusChanged(true);
+                  }
+                  break;
+              default:
+                  break;
+          }
+          final SimpleExoPlayer player = refPlayer.get();
+          if (player != null) {
+              if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK) {
+                  // Lower the volume
+                  player.setVolume(mAudioVolume * 0.8f);
+              } else if (focusChange == AudioManager.AUDIOFOCUS_GAIN) {
+                  // Raise it back to normal
+                  player.setVolume(mAudioVolume * 1);
+              }
+          }
+      }
+  }
+
 }

**Comment by Tang-xun** [comment]
###  It's my solution , maybe helpful to you .

````java
Index: react-native-video/android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- node_modules/react-native-video/android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java	(revision e4478836f33397e8479bf24a51d83db5514b728f)
+++ node_modules/react-native-video/android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java	(revision a78bc362053c37f43f5058020079579041820914)
@@ -67,6 +67,7 @@
 import com.google.android.exoplayer2.ui.PlayerControlView;
 import com.google.android.exoplayer2.video.VideoFrameMetadataListener;
 
+import java.lang.ref.WeakReference;
 import java.net.CookieHandler;
 import java.net.CookieManager;
 import java.net.CookiePolicy;
@@ -82,7 +83,6 @@
         ExoPlayer.EventListener,
         BandwidthMeter.EventListener,
         BecomingNoisyListener,
-        AudioManager.OnAudioFocusChangeListener,
         VideoFrameMetadataListener,
         MetadataRenderer.Output {
 
@@ -102,7 +102,7 @@
     private PlayerControlView playerControlView;
     private View playPauseControlContainer;
     private Player.EventListener eventListener;
-    
+
     private Handler mainHandler;
     private ExoPlayerView exoPlayerView;
 
@@ -137,7 +137,7 @@
     private String audioTrackType;
     private Dynamic audioTrackValue;
     private String videoTrackType;
-    private Dynamic videoTrackValue;    
+    private Dynamic videoTrackValue;
     private ReadableArray audioTracks;
     private String textTrackType;
     private Dynamic textTrackValue;
@@ -175,13 +175,15 @@
         }
     };
 
+    private ExoOnAudioFocusChangeListener mOnAudioFocusChangeListener;
+
     public ReactExoplayerView(ThemedReactContext context) {
         super(context);
         this.themedReactContext = context;
         this.eventEmitter = new VideoEventEmitter(context);
 
         createViews();
-        
+
         audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
         themedReactContext.addLifecycleEventListener(this);
         audioBecomingNoisyReceiver = new AudioBecomingNoisyReceiver(themedReactContext);
@@ -466,14 +468,20 @@
         themedReactContext.removeLifecycleEventListener(this);
         audioBecomingNoisyReceiver.removeListener();
         BANDWIDTH_METER.removeEventListener(this);
-        audioManager.abandonAudioFocus(this);
+        audioManager.abandonAudioFocus(mOnAudioFocusChangeListener);
     }
 
     private boolean requestAudioFocus() {
         if (disableFocus || srcUri == null) {
             return true;
         }
-        int result = audioManager.requestAudioFocus(this,
+
+        if (mOnAudioFocusChangeListener == null) {
+            mOnAudioFocusChangeListener = new ExoOnAudioFocusChangeListener(player, eventEmitter);
+            updateAudioFocusChangeVolume();
+        }
+
+        int result = audioManager.requestAudioFocus(mOnAudioFocusChangeListener,
                 AudioManager.STREAM_MUSIC,
                 AudioManager.AUDIOFOCUS_GAIN);
         return result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED;
@@ -538,7 +546,7 @@
             setFullscreen(false);
         }
         setKeepScreenOn(false);
-        audioManager.abandonAudioFocus(this);
+        audioManager.abandonAudioFocus(mOnAudioFocusChangeListener);
     }
 
     private void updateResumePosition() {
@@ -565,29 +573,29 @@
 
     // AudioManager.OnAudioFocusChangeListener implementation
 
-    @Override
-    public void onAudioFocusChange(int focusChange) {
-        switch (focusChange) {
-            case AudioManager.AUDIOFOCUS_LOSS:
-                eventEmitter.audioFocusChanged(false);
-                break;
-            case AudioManager.AUDIOFOCUS_GAIN:
-                eventEmitter.audioFocusChanged(true);
-                break;
-            default:
-                break;
-        }
-
-        if (player != null) {
-            if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK) {
-                // Lower the volume
-                player.setVolume(audioVolume * 0.8f);
-            } else if (focusChange == AudioManager.AUDIOFOCUS_GAIN) {
-                // Raise it back to normal
-                player.setVolume(audioVolume * 1);
-            }
-        }
-    }
+//    @Override
+//    public void onAudioFocusChange(int focusChange) {
+//        switch (focusChange) {
+//            case AudioManager.AUDIOFOCUS_LOSS:
+//                eventEmitter.audioFocusChanged(false);
+//                break;
+//            case AudioManager.AUDIOFOCUS_GAIN:
+//                eventEmitter.audioFocusChanged(true);
+//                break;
+//            default:
+//                break;
+//        }
+//
+//        if (player != null) {
+//            if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK) {
+//                // Lower the volume
+//                player.setVolume(audioVolume * 0.8f);
+//            } else if (focusChange == AudioManager.AUDIOFOCUS_GAIN) {
+//                // Raise it back to normal
+//                player.setVolume(audioVolume * 1);
+//            }
+//        }
+//    }
 
     // AudioBecomingNoisyListener implementation
 
@@ -884,7 +892,7 @@
 
     public void setReportBandwidth(boolean reportBandwidth) {
         mReportBandwidth = reportBandwidth;
-    }   
+    }
 
     public void setRawSrc(final Uri uri, final String extension) {
         if (uri != null) {
@@ -1009,7 +1017,7 @@
             for (int j = 0; j < group.length; j++) {
                 tracks[j] = j;
             }
-        } 
+        }
 
         if (groupIndex == C.INDEX_UNSET) {
             trackSelector.setParameters(disableParameters);
@@ -1078,6 +1086,7 @@
         if (player != null) {
             player.setVolume(audioVolume);
         }
+        updateAudioFocusChangeVolume();
     }
 
 
@@ -1086,6 +1095,13 @@
         if (player != null) {
             player.setVolume(audioVolume);
         }
+        updateAudioFocusChangeVolume();
+    }
+
+    private void updateAudioFocusChangeVolume() {
+        if (mOnAudioFocusChangeListener != null) {
+            mOnAudioFocusChangeListener.setAudioVolume(audioVolume);
+        }
     }
 
     public void seekTo(long positionMs) {
@@ -1178,7 +1194,7 @@
 
     /**
      * Handling controls prop
-     * 
+     *
      * @param controls  Controls prop, if true enable controls, if false disable them
      */
     public void setControls(boolean controls) {
@@ -1198,4 +1214,53 @@
   public void onVideoFrameAboutToBeRendered(long presentationTimeUs, long releaseTimeNs, Format format) {
     this.presentationTimeUs = presentationTimeUs;
   }
+
+  // fix leaks for audio focus change listener
+  static class ExoOnAudioFocusChangeListener
+      implements AudioManager.OnAudioFocusChangeListener {
+
+
+      WeakReference<SimpleExoPlayer> refPlayer;
+      WeakReference<VideoEventEmitter> refEventEmitter;
+
+      float mAudioVolume;
+
+      public ExoOnAudioFocusChangeListener(SimpleExoPlayer player, VideoEventEmitter eventEmitter) {
+          refEventEmitter = new WeakReference<>(eventEmitter);
+          refPlayer = new WeakReference<>(player);
+      }
+
+      public void setAudioVolume(float audioVolume) {
+          mAudioVolume = audioVolume;
+      }
+
+      @Override
+      public void onAudioFocusChange(int focusChange) {
+          switch (focusChange) {
+              case AudioManager.AUDIOFOCUS_LOSS:
+                  if (refEventEmitter.get() != null) {
+                      refEventEmitter.get().audioFocusChanged(false);
+                  }
+                  break;
+              case AudioManager.AUDIOFOCUS_GAIN:
+                  if (refEventEmitter.get() != null) {
+                      refEventEmitter.get().audioFocusChanged(true);
+                  }
+                  break;
+              default:
+                  break;
+          }
+          final SimpleExoPlayer player = refPlayer.get();
+          if (player != null) {
+              if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK) {
+                  // Lower the volume
+                  player.setVolume(mAudioVolume * 0.8f);
+              } else if (focusChange == AudioManager.AUDIOFOCUS_GAIN) {
+                  // Raise it back to normal
+                  player.setVolume(mAudioVolume * 1);
+              }
+          }
+      }
+  }
+
 }


````

**Comment by BasixKOR** [comment]
I can confirm this issue is still happening on the latest 5.2.x line.

**Comment by freeboub** [comment]
linked PR merge, will be available in 6.0.0-alpha.7
